woocommerce/woocommerce-ios
# Swift 6 Migration: Dynamic dashboard

Part of #13001


## Description

This PR updates the dynamic dashboard feature to support the `targeted` mode for concurrency checking. This supports early preparation for Swift 6 migration.

To kick-start the migration, I marked the `DashboardViewModel` as main actor-isolated to make it safe to use for the `DashboardView`. With the `targeted` mode for concurrency checking set, Xcode then threw errors and warnings regarding this change. I addressed the issues following the rules:
- If a type is a model to be used in async contexts (sent from the Networking layer to the UI), mark it as `Sendable`.
- If a type is an observable object to be used as a view model, mark it as main actor-isolated to be safely used for the UI on the main thread.
- If a method makes a call to a main actor-isolated initializer/method in a synchronous nonisolated context:
  - Mark it as `@MainActor` if possible.
  - In cases where the method is a protocol-conformance, Xcode would require marking the method in the protocol `@MainActor`. This would affect more places in the app and make the scope of change much larger depending on how many models conform to this same protocol. A quick fix would be to wrap the implementation of the method in a main actor-isolated `Task`.
- Tests that make calls to main actor-isolated initializers needs to mark with `@MainActor` as well. 

## Testing steps

- In Xcode settings, set `SWIFT_STRICT_CONCURRENCY` to `targeted` and build the app. Confirm that the app builds successfully and runs normally.
- Revert the change in Xcode settings and confirm that the app still works correctly with the default setting of `SWIFT_STRICT_CONCURRENCY`.
- Unit tests should build successfully and pass.

## Screenshots

N/A

---
